### PR TITLE
Issue 19563 - non-POD C++ argument not handled correctly

### PR DESCRIPTION
**Pull request by SSoulaimane**
According to the System V 64 bits ABI non-POD function arguments should be passed by pointer.
> Section 3.2.3 page 21:
> 2. If a C++ object is non-trivial for the purpose of calls, as specified in the
C++ ABI , it is passed by invisible reference (the object is replaced in the
parameter list by a pointer that has class INTEGER).

### CODE REVIEW

File: src/dmd/toir.d [hunk 1]
```diff
@@ -1058,8 +1058,8 @@ void buildClosure(FuncDeclaration fd, IRState *irs)
             if (!v.isParameter())
                 continue;
             tym_t tym = totym(v.type);
-            bool win64ref = ISWIN64REF(v);
-            if (win64ref)
+            bool x64ref = ISX64REF(v);
```
Comment (jacob-carlborg) [inline src/dmd/toir.d]: `const`.

**Comment by thewilsonator** [comment]
You need to add "fix" to the title of one of your commits.

File: src/dmd/e2ir.d [hunk 2]
```diff
@@ -89,17 +89,19 @@ void* mem_malloc2(uint);
  */
 bool ISREF(Declaration var)
 {
-    return (config.exe == EX_WIN64 && var.isParameter() &&
-            (var.type.size(Loc.initial) > REGSIZE || var.storage_class & STC.lazy_))
+    return (var.isParameter() &&
+            ((!global.params.isWindows && var.type.isTypeStruct() && !var.type.isTypeStruct().sym.isPOD()) ||
```
Comment (WalterBright) [inline src/dmd/e2ir.d]: Having things like `!global.params.isWindows` is a future bug lying in wait. It assumes that all future systems supported by the compiler that are not Windows will conform to this ABI.
It's far batter to enumerate which targets it applies to, which is what `config.exe` is for, as it is a bit mask.

**Comment by WalterBright** [comment]
> Section 3.2.3 page 21:
2. If a C++ object is non-trivial for the purpose of calls, as specified in the
C++ ABI , it is passed by invisible reference (the object is replaced in the
parameter list by a pointer that has class INTEGER).

This should appear as a comment in the code in the appropriate place.

**Comment by WalterBright** [comment]
This is a 64 bit ABI change, but the 32 bit tests are all failing as well. So something has gone awry with the patch.

**Comment by SSoulaimane** [comment]
> This is a 64 bit ABI change, but the 32 bit tests are all failing as well. So something has gone awry with the patch.

No, I changed it to include 32 bits posix as well since I found it has the same behavior.

**Comment by look-at-me** [comment]
There's another related bug, that this also won't take into consideration. The `isPOD` doesn't actually correspond to C++'s definition of it. If you have a default constructor in C++ then it is no POD. There is no equivalent in D cause there are no default constructors.

```C++
#include <iostream>
#include <type_traits>

struct V {
	float x = 0.0f;
	float y = 0.0f;
};

struct X {
	float x;
	float y;

	X() : x(0), y(0)
	{
	}
};

int main() {
	std::cout << std::is_pod<V>::value; // 0 (false)
	std::cout << std::is_pod<X>::value; // 0 (false)
	return 0;
}
```

**Comment by SSoulaimane** [comment]
> ...There is no equivalent in D cause there are no default constructors.
It's a language thing. people usually get around it by defining a postblit on the D side.

**Comment by look-at-me** [comment]
Yes that adds bloat to an otherwise simple type. For example a Vector2, if you use it with containers that check if the type has a postblit, it will see it does and add in extra copying logic for something that would just be a memcpy() all to satisfy an ABI issue with C++.

**Comment by WalterBright** [comment]
Can the failures be fixed?

**Pull request closed without merging**